### PR TITLE
Add support for Teensy-modified Datahand keyboard

### PR DESCRIPTION
**Pull request by adzenith**
## Description
I have a DataHand keyboard in which I've replaced the stock processor with a [Teensy++ 2.0](https://www.pjrc.com/store/teensypp.html) in a similar manner to that described [here](https://geekhack.org/index.php?topic=12212.0). This PR adds a new `keyboard` to support this setup.

## Types of changes
- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [ ] Enhancement/Optimization
- [X] Keyboard (addition or update)
- [X] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Checklist:
- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/datahand/datahand.h [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+// This a shortcut to help you visually see your layout.
+// The first section contains all of the arguements; the second converts the arguments into a two-dimensional array.
+
+// Each code is three letters
+// l or r - left or right hand
+// p, r, m, i, t - pinky, ring finger, middle finger, index finger, thumb
+// fingers: n, s, e, w, c - north, south, east, west, and center (manual calls this "well" but we already have "west")
+// thumb:   p, n, c, l, u, k - pad, nail, center, lock (harder center), up, knuckle
+#define KEYMAP( \
```
[the remark this answers is not in the document]
Comment (adzenith) [reply keyboards/datahand/datahand.h]: Gotcha - I'll add `KC_` everywhere.

**Comment by adzenith** [comment]
Thanks for the quick review! I've updated the PR as per your suggestions: adding `KC_` everywhere, fixing header guards, updating to `process_user_record`, etc.

File: quantum/matrix.c [hunk 2]
```diff
@@ -63,9 +63,9 @@ static const pin_t col_pins[MATRIX_COLS] = MATRIX_COL_PINS;
 #endif
 
 /* matrix state(1:on, 0:off) */
-static matrix_row_t matrix[MATRIX_ROWS];
+__attribute__((weak)) matrix_row_t matrix[MATRIX_ROWS];
```
Comment (drashna) [inline quantum/matrix.c]: These files shouldn't be modified, and instead, you should be using a custom matrix.

File: tmk_core/common/action_layer.c [hunk 3]
```diff
@@ -140,7 +140,7 @@ void layer_state_set(uint32_t state)
 
 /** \brief Layer clear
  *
- * FIXME: Needs docs
+ * Turn off all layers.
```
Comment (drashna) [inline tmk_core/common/action_layer.c]: It's great to see this, but could these changes be moved to a separate PR?

Comment (adzenith) [reply tmk_core/common/action_layer.c]: Definitely!

File: quantum/matrix.c (hunk 2 above)
Comment (adzenith) [reply quantum/matrix.c]: I am using a custom matrix, but I wanted to use functions from this file without having a second `matrix` variable linked in. I'll take this change out and copy the functions from this file to my `matrix.c`.

**Comment by adzenith** [comment]
I'll move it to `handwired`. Thanks for the review!

**Comment by adzenith** [comment]
Updated!

File: tmk_core/common/action_layer.c (hunk 3 above)
Comment (adzenith) [reply tmk_core/common/action_layer.c]: https://github.com/qmk/qmk_firmware/pull/4858

File: keyboards/handwired/datahand/rules.mk [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# Project-specific includes
+SRC = matrix.c
+
+# MCU name
+MCU = AT90USB1286
```
Comment (drashna) [inline keyboards/handwired/datahand/rules.mk]: The travis CI error is occurring because of this, I think.

This should be lower case, and yes that does matter.

```suggestion
MCU = at90usb1286
```

**Comment by adzenith** [comment]
Looks like CI is working now! Let me know if there's anything else I need to do. Thanks again for the help!

**Comment by drashna** [comment]
This is more of a nit pick, but could you update the code (especially the default keymap) to conform to the QMK Coding conventions? 
https://docs.qmk.fm/#/contributing?id=coding-conventions

**Comment by adzenith** [comment]
Sure! I can change the comments to use C-style comments (which it says is "strongly encouraged") – other than that, I think I already am complying with coding standards. Is there anything else I'm missing?
Thanks!

**Comment by adzenith** [comment]
Updated with C-style comments.

**Comment by drashna** [comment]
The comments are fine. 

Mostly: https://github.com/qmk/qmk_firmware/blob/df34c46d1f939ff37fcf24daf94800c65fdeceff/keyboards/handwired/datahand/keymaps/default/keymap.c#L137-L238

**Comment by adzenith** [comment]
Hm, I'm not sure what you mean. I'm looking at the coding convention and I see

* We indent using two spaces (soft tabs)
* We use a modified One True Brace Style
* We encourage use of C style comments: /* */
* In general we don't wrap lines
* We use #pragma once

I do see a couple of misplaced braces in keymap.c, which I can fix, but I'm not sure what else is being pointing out in those sections.
Are you referring to the function opening braces? I can move those onto the same lines as the function declarations if so. The coding standard says block braces have to be on the same line, but function braces I wasn't sure about because a different style is used in e.g. `quantum/matrix.c` and `quantum/quantum.c`

**Comment by adzenith** [comment]
I've moved the function braces onto the function lines - let me know if that's what you meant!

**Comment by drashna** [comment]
Sorry for not being clear, yeah that's what I meant!

**Comment by drashna** [comment]
Thanks!
